### PR TITLE
OHRM5X-988: Timesheet Create operation added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following changes:
1. Timesheet Endpoint added.
2. Timesheet Dao and Service layer moved from lib.
3. Data group permission and role permission added for the end point.
4. API, DAO test added.
5. Unique Validation added for checking the duplicate entry for start date.

### CODE REVIEW

File: symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+use OrangeHRM\Core\Api\V2\Validator\ParamRuleCollection;
+use OrangeHRM\Core\Api\V2\Validator\Rule;
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+use OrangeHRM\Core\Traits\Auth\AuthUserTrait;
+use OrangeHRM\Core\Traits\Service\DateTimeHelperTrait;
+use OrangeHRM\Entity\Timesheet;
+use OrangeHRM\Time\Api\Model\TimesheetModel;
+use OrangeHRM\Time\Traits\Service\TimesheetServiceTrait;
+
+class TimesheetAPI extends Endpoint implements CollectionEndpoint
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php:37]:
```suggestion
class MyTimesheetAPI extends Endpoint implements CollectionEndpoint
```
shall we change?

File: symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    new Rule(Rules::API_DATE),
+                    new Rule(Rules::CALLBACK, [
+                        function () {
+                            $date = $this->getRequestParams()->getDateTime(
+                                RequestParams::PARAM_TYPE_BODY,
+                                self::PARAMETER_DATE
+                            );
+                            return !($date > $this->getDateTimeHelper()->getNow());
+                        }
+                    ])
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php:92]: Better to handle if timesheet has already created

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 3]
```diff
@@ -71,24 +79,22 @@ public function getTimesheetById($timesheetId) {
...
-     * @param TimesheetDao $TimesheetDao
+     * @param TimesheetDao $timesheetDao
      * @return void
      */
-    public function setTimesheetDao(TimesheetDao $timesheetDao) {
-
+    public function setTimesheetDao(TimesheetDao $timesheetDao) :void
+    {
         $this->timesheetDao = $timesheetDao;
     }
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:99]: do we need this?

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 4]
```diff
@@ -456,12 +465,28 @@ public function checkForOverlappingTimesheets($startDate, $endDate, $employeeId)
...
     /**
-     * Add, Update Timesheet
      * @param Timesheet $timesheet
-     * @return boolean
+     * @param DateTime $date
+     * @return Timesheet
      */
-    public function saveTimesheet(Timesheet $timesheet) {
-
+    public function saveTimesheet(Timesheet $timesheet, DateTime $date): Timesheet
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:472]:
```suggestion
    public function createTimesheetByDate(Timesheet $timesheet, DateTime $date): Timesheet
```
what do you think?

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 5]
```diff
@@ -456,12 +465,28 @@ public function checkForOverlappingTimesheets($startDate, $endDate, $employeeId)
...
+    {
+        $accessFlowStateMachineService = new AccessFlowStateMachineService();
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:474]: this service initialization you can move to getter method in this class

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 6]
```diff
@@ -456,12 +465,28 @@ public function checkForOverlappingTimesheets($startDate, $endDate, $employeeId)
...
+        $tempNextState = $accessFlowStateMachineService->getNextState(
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:475]:
```suggestion
        $nextState = $accessFlowStateMachineService->getNextState(
```

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 7]
```diff
@@ -42,6 +42,13 @@ apiv2_time_project_activity:
   defaults:
     _api: OrangeHRM\Time\Api\ProjectActivityAPI
 
+apiv2_time_my_time_sheet:
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/config/routes.yaml:45]: :+1:

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml [hunk 8]
```diff
@@ -42,6 +42,13 @@ apiv2_time_project_activity:
...
+  path: /api/v2/time/timesheets
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ POST ]
+  defaults:
+    _api: OrangeHRM\Time\Api\TimesheetAPI
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/config/routes.yaml:50]:
```suggestion
    _api: OrangeHRM\Time\Api\MyTimesheetAPI
```

File: symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function getStartDate(): ?string
+    {
+        return $this->getDateTimeHelper()->formatDateTimeToYmd($this->getTimesheet()->getStartDate());
+    }
+
+    /**
+     * @return Timesheet
+     */
+    public function getTimesheet(): Timesheet
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php:50]:
```suggestion
    protected function getTimesheet(): Timesheet
```

File: symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function __construct(Timesheet $timesheet)
+    {
+        $this->timesheet = $timesheet;
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getStartDate(): ?string
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php:42]:
```suggestion
    /**
     * @return string
     */
    public function getStartDate(): string
```
Hope startDate and endDate non null values in Timesheet entity

File: symfony/plugins/orangehrmTimePlugin/test/fixtures/ProjectActivityDao.yml [hunk 11]
```diff
@@ -1,5 +1,10 @@
 Config:
-  - { name: 'authorize_user_role_manager_class', value: 'BasicUserRoleManager' }
+  0: { name: 'authorize_user_role_manager_class', value: 'BasicUserRoleManager' }
+  1: { name: 'timesheet_period_set', value: 'Yes' }
+  2: { name: 'timesheet_period_and_start_date', value: '<TimesheetPeriod><PeriodType>Weekly</PeriodType><ClassName>WeeklyTimesheetPeriod</ClassName><StartDate>1</StartDate><Heading>Week</Heading></TimesheetPeriod>' }
+
+WorkflowStateMachine:
+  - { "id": "1", "workflow": "0", "state": "INITIAL", "role": "SYSTEM", "action": "7", "resulting_state": "NOT SUBMITTED", "priority": 0 }
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/test/fixtures/ProjectActivityDao.yml:7]: *

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/dbscript/dbscript-2.sql [hunk 12]
```diff
@@ -4244,26 +4244,31 @@ INSERT INTO ohrm_data_group (`name`, `description`, `can_read`, `can_create`, `c
...
 SET @apiv2_project_activities_data_group_id := (SELECT `id` FROM ohrm_data_group WHERE name = 'apiv2_project_activities' LIMIT 1);
+SET @apiv2_time_timesheet_data_group_id := (SELECT `id` FROM ohrm_data_group WHERE name = 'apiv2_time_timesheet' LIMIT 1);
 
 INSERT INTO ohrm_api_permission (`api_name`, `module_id`, `data_group_id`)
 VALUES ('OrangeHRM\\Time\\Api\\TimeConfigPeriodAPI', @time_module_id, @apiv2_time_time_sheet_config_data_group_id),
        ('OrangeHRM\\Time\\Api\\CustomerAPI', @time_module_id, @apiv2_time_customers_data_group_id),
        ('OrangeHRM\\Time\\Api\\ValidationCustomerNameAPI', @time_module_id, @apiv2_time_validation_customer_name_data_group_id),
-       ('OrangeHRM\\Time\\Api\\ProjectActivityAPI', @time_module_id, @apiv2_project_activities_data_group_id);
+       ('OrangeHRM\\Time\\Api\\ProjectActivityAPI', @time_module_id, @apiv2_project_activities_data_group_id),
+       ('OrangeHRM\\Time\\Api\\TimesheetAPI', @time_module_id, @apiv2_time_timesheet_data_group_id);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply installer/dbscript/dbscript-2.sql]: done

File: symfony/lib/framework/Services.php [hunk 13]
```diff
@@ -240,4 +240,9 @@ final class Services
      * @see \OrangeHRM\Time\Service\CustomerService
      */
     public const CUSTOMER_SERVICE = 'time.customer_service';
+
+    /**
+     * @see \OrangeHRM\Time\Service\CustomerService
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/lib/framework/Services.php]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php (hunk 3 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml (hunk 7 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml:68]: done

File: symfony/plugins/orangehrmTimePlugin/config/routes.yaml (hunk 8 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/config/routes.yaml]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php (hunk 4 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php (hunk 5 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php (hunk 6 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php]: done

File: symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php (hunk 9 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php]: done

File: symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php (hunk 10 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/entity/Decorator/TimesheetDecorator.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php (hunk 1 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php (hunk 2 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/TimesheetAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/test/fixtures/ProjectActivityDao.yml (hunk 11 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/test/fixtures/ProjectActivityDao.yml]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 14]
```diff
@@ -503,4 +521,47 @@ protected function getTimesheetData(Timesheet $timesheet): array
...
+        $timesheet->setStartDate(new DateTime($startDate));
+        $timesheet->setEndDate(new DateTime($endDate));
+        return $this->getTimesheetDao()->saveTimesheet($timesheet);
+    }
+
+    /**
+     * @param DateTime $date
+     * @return array
+     */
+    public function extractStartDateAndEndDateFromDate(DateTime $date): array
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:549]: shall we write more unit tests, by mocking `$this->getTimeSheetPeriodService()->getTimesheetStartDate()`

**Review by RajithaKumara — APPROVED** [review]

File: symfony/plugins/orangehrmTimePlugin/Api/MyTimesheetAPI.php [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                    new Rule(MyTimesheetDateRule::class),
+                    new Rule(Rules::CALLBACK, [
+                        function () {
+                            $date = $this->getRequestParams()->getDateTime(
+                                RequestParams::PARAM_TYPE_BODY,
+                                self::PARAMETER_DATE
+                            );
+                            return !($date > $this->getDateTimeHelper()->getNow());
+                        }
+                    ])
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/MyTimesheetAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 16]
```diff
@@ -820,4 +816,16 @@ public function checkForMatchingTimesheetForCurrentDate($employeeId, $currentDat
             return null;
         }
     }
+
+    /**
+     * @param DateTime $date
+     * @return bool
+     */
+    public function isTimesheetTakenByDate(DateTime $date): bool
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 17]
```diff
@@ -64,6 +72,17 @@ public function setTimesheetPeriodDao(TimesheetPeriodService $timesheetPeriodSer
         $this->timesheetPeriodService = $timesheetPeriodService;
     }
 
+    /**
+     * @return AccessFlowStateMachineService
+     */
+    public function getAccessFlowStateMachineService(): AccessFlowStateMachineService
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php]: done

File: symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php [hunk 18]
```diff
@@ -503,4 +521,47 @@ protected function getTimesheetData(Timesheet $timesheet): array
...
+        $startDate = date('Y-m-d', strtotime($currentWeekFirstDate . ' + ' . $configDate . ' days'));
+        $endDate = date('Y-m-d', strtotime($startDate . ' + 6 days'));
+        return [$startDate, $endDate];
+    }
+
+    /**
+     * @param DateTime $date
+     * @return bool
+     */
+    public function isTimesheetTakenByDate(DateTime $date)
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:562]: done

[the remark this answers is not in the document]
Comment (RajithaKumara) [reply symfony/plugins/orangehrmTimePlugin/Service/TimesheetService.php:562]: Return type also not defined

**Pull request merged**